gostream/ffmpeg/avcodec.go

```diff
@@ -0,0 +1,679 @@
+	CodecContext struct {
+		p  *C.struct_AVCodecContext
+		pp **C.struct_AVCodecContext
+	}
+
+	// Packet an AVPacket
+	Packet struct {
+		p  *C.struct_AVPacket
+		pp **C.struct_AVPacket
+	}
+	// PixelFormat an AVPixelFormat
+	PixelFormat C.enum_AVPixelFormat
+	// CodecID an AVCodecID
+	CodecID C.enum_AVCodecID
+)
+
+// From https://ffmpeg.org/doxygen/4.0/avcodec_8h_source.html#l00215
+const (
+	CodecIDNone = iota
+	CodecIDMPEG1VIDEO
+	CodecIDMPEG2VIDEO
+	CodecIDH261
+	CodecIDH263
+	CodecIDRV10
+	CodecIDRV20
+	CodecIDMJPEG
+	CodecIDMJPEGB
+	CodecIDLJPEG
+	CodecIDSP5X
+	CodecIDJPEGLS
+	CodecIDMPEG4
+	CodecIDRAWVIDEO
+	CodecIDMSMPEG4V1
+	CodecIDMSMPEG4V2
+	CodecIDMSMPEG4V3
+	CodecIDWMV1
+	CodecIDWMV2
+	CodecIDH263P
+	CodecIDH263I
+	CodecIDFLV1
+	CodecIDSVQ1
+	CodecIDSVQ3
+	CodecIDDVVIDEO
+	CodecIDHUFFYUV
+	CodecIDCYUV
+	CodecIDH264
+	CodecIDINDEO3
+	CodecIDVP3
+)
+
+// FindEncoderByName Find a registered encoder with the specified name.
+//
+// @param name the name of the requested encoder
+// @return An encoder if one was found, NULL otherwise.
+func FindEncoderByName(c string) (Codec, error) {
+	codec := C.avcodec_find_encoder_by_name(C.CString(c))
+	if codec == nil {
+		return Codec{}, errors.Errorf("cannot find encoder %s", c)
+	}
+	return Codec{
+		p:  codec,
+		pp: &codec,
+	}, nil
+}
+
+// ToString get the string representation of the codec ID
+func (c *CodecID) ToString() string {
+	switch *c {
+	case CodecIDNone:
+		return "None"
+	case CodecIDMPEG1VIDEO:
+		return "MPEG1VIDEO"
+	case CodecIDMPEG2VIDEO:
+		return "MPEG2VIDEO"
+	case CodecIDH261:
+		return "H261"
+	case CodecIDH263:
+		return "H263"
+	case CodecIDRV10:
+		return "RV10"
+	case CodecIDRV20:
+		return "RV20"
+	case CodecIDMJPEG:
+		return "MJPEG"
+	case CodecIDMJPEGB:
+		return "MJPEGB"
+	case CodecIDLJPEG:
+		return "LJPEG"
+	case CodecIDSP5X:
+		return "SP5X"
+	case CodecIDJPEGLS:
+		return "JPEGLS"
+	case CodecIDMPEG4:
+		return "MPEG4"
+	case CodecIDRAWVIDEO:
+		return "RAWVIDEO"
+	case CodecIDMSMPEG4V1:
+		return "MSMPEG4V1"
+	case CodecIDMSMPEG4V2:
+		return "MSMPEG4V2"
+	case CodecIDMSMPEG4V3:
+		return "MSMPEG4V3"
+	case CodecIDWMV1:
+		return "WMV1"
+	case CodecIDWMV2:
+		return "WMV2"
+	case CodecIDH263P:
+		return "H263P"
+	case CodecIDH263I:
+		return "H263I"
+	case CodecIDFLV1:
+		return "FLV1"
+	case CodecIDSVQ1:
+		return "SVQ1"
+	case CodecIDSVQ3:
+		return "SVQ3"
+	case CodecIDDVVIDEO:
+		return "DVVIDEO"
+	case CodecIDHUFFYUV:
+		return "HUFFYUV"
+	case CodecIDCYUV:
+		return "CYUV"
+	case CodecIDH264:
+		return "H264"
+	case CodecIDINDEO3:
+		return "INDEO3"
+	case CodecIDVP3:
+		return "VP3"
+	default:
+		return "unknown"
+	}
+}
+
+// FindDecoder Find a registered decoder with a matching codec ID.
+//
+// @param id AVCodecID of the requested decoder
+// @return A decoder if one was found, NULL otherwise.
+func FindDecoder(ctx context.Context, id CodecID) (Codec, error) {
+	var zero Codec
+	select {
+	case <-ctx.Done():
+		return zero, ctx.Err()
```

Comment:
Added context so these methods can fail fast the moment the context is canceled. To cancel sooner than this there's a callback we can use https://ffmpeg.org/doxygen/trunk/structAVFormatContext.html#a5b37acfe4024d92ee510064e80920b40